resilience4j-micronaut/src/main/java/io/github/resilience4j/mirconaut/BaseInterceptor.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 Michael Pollind
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.mirconaut;
+
+import io.github.resilience4j.bulkhead.operator.BulkheadOperator;
+import io.github.resilience4j.mirconaut.fallback.UnhandledFallbackException;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.discovery.exceptions.NoAvailableServiceException;
+import io.micronaut.inject.MethodExecutionHandle;
+import io.reactivex.Flowable;
+import io.reactivex.FlowableTransformer;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+public abstract class BaseInterceptor {
```

Comment:
For this, you may want to consider in Micronaut 2.1 using the new `InterceptedMethod` API which simplifies handling future/reactive/synchronous cases. See https://github.com/micronaut-projects/micronaut-core/blob/2.1.x/runtime/src/main/java/io/micronaut/retry/intercept/RecoveryInterceptor.java#L79